synapse/storage/databases/main/devices.py

```diff
@@ -706,7 +709,7 @@ async def get_user_devices_from_cache(
                 device = await self._get_cached_user_device(user_id, device_id)
                 results.setdefault(user_id, {})[device_id] = device
             else:
-                results[user_id] = await self.get_cached_devices_for_user(user_id)
+                results[user_id] = dict(await self.get_cached_devices_for_user(user_id))
```

Comment:
`get_cached_devices_for_user` returns a cached result.

We can't guarantee that the branch above won't be taken later on, so we have to take a copy here. Which is unfortunate.

Reply:
This code seems kind of confusing in the first place -- it doesn't make sense to have a user with device IDs & all devices. I think some refactoring will help.

Reply:
#15040.